FRA Error Report

elipe17 opened the pull request:
## Summary of Changes
- Added support for FRA error report generation
- Re-factored report generation into class based structure
- Added unit tests for report download

Pull request closes #3419 

## How to Test

```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Submit some FRA files
3. Download report from submission history link if it's been merged... Otherwise use the Django shell to generate the file

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [ ] TDP Users should be able to download error reports for FRA file types
- [ ] Error report spreadsheets should closely mimic Mock ups seen under Supporting Documentation
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Allow ofa sys admin to submit fra without the flag
- Allow frontend to submit frontend files and kick off parsing chain

elipe17 committed:
- Add new error reports file
- Add class based structure for error reports
- Create factory for error report classes
- Remove old functional erorr reports code
- Leverage error report generator in datafile view
- Left TODO on rpt_month_year
- Update FRA schema to have item numbers

elipe17 committed:
- Linting

elipe17 committed:
- Remove exit date form parser error
- Bring all TE1 records into memory for the moment

elipe17 committed:
- add todo

elipe17 committed:
- Update FRA report to use raw_sql queries since we need to access the model's field data

elipe17 committed:
- use docstring for convenience

elipe17 committed:
- Use good names for query

elipe17 committed:
- Close fra workbook
- Add tests for FRA file error reports

elipe17 committed:
Merge branch '3474-fra-parser' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
Merge branch '3474-fra-parser' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
Merge branch '3474-fra-parser' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
Merge branch '3474-fra-parser' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
- Update to serialize record to db even though we are generating cat4 errors
- Update tests to reflect new behavior

elipe17 committed:
Merge branch '3474-fra-parser' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
Merge branch '3474-fra-parser' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 commented on tdrs-backend/tdpservice/data_files/error_reports.py:
```diff
@@ -0,0 +1,353 @@
+                                           pe.object_id, pe.error_message, pe.fields_json,
+                                           pe.field_name, pe.row_number, pe.column_number
+                                           FROM search_indexes_tanf_exiter1 AS te1
+                                           INNER JOIN parser_error pe ON te1.id = pe.object_id""")
```
Add extra join on datafile

raftmsohani replied:
you should be able to to use ```.prefetch_related``` and not have to do raw query. There are cases that raw query makes more sense, but this is only a simple join

jtimpe replied:
i was also curious if `prefetch_related` or something similar might help here. prefer not to have raw SQL if we can avoid it!

raftmsohani commented on tdrs-backend/tdpservice/data_files/error_reports.py:
```diff
@@ -0,0 +1,353 @@
+    @staticmethod
+    def get_error_report_generator(datafile):
+        """Get error report generator."""
+        active = DataFile.Section.ACTIVE_CASE_DATA
```
nit: active -> datafile_active_case

raftmsohani commented on tdrs-backend/tdpservice/data_files/test/test_api.py:
```diff
@@ -39,6 +39,26 @@ def test_datafile(self, stt_user, stt):
+                                                  DataFile.Section.FRA_WORK_OUTCOME_TANF_EXITERS)
+        test_datafile.year = 2024
+        test_datafile.quarter = 'Q1'
+        test_datafile.save()
```
I don't think we need to save the datafile since we are returning the object

elipe17 committed:
Merge branch '3474-fra-parser' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
Merge branch '3527-parser-error-values' into 3419-fra-error-report

elipe17 committed:
- Update cat1 and cat4 errors to contian refs to all fields and values for the offending record.
- Remove refs to field in cat4 validator class
- Add logic to util.py to maintain existing file status checks wrt field
- remove print form test

elipe17 committed:
- Update preparsing errors to be associated with a record
- Update tests to reflect records not being serialized anymore

elipe17 committed:
- correct tdr schema to return correct type

elipe17 committed:
- Update fra error report to leverage more informed parser error objects
- update api error report tests to refelect new test cases
- for alpha fields to be strings
- Create custom error generator for fra
- Override run_preparsing_validators for fra

elipe17 committed:
- Update fra parser to leverage fra specific error generator
- Add check for precheck error to null out record since we're done with it at taht point
- Update tests to take a null record

elipe17 committed:
Merge branch '3527-parser-error-values' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
- linting

elipe17 committed:
- remove comment

elipe17 commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -48,7 +49,13 @@ def generate_parser_error(datafile, line_number, schema, error_category, error_m
+    # result field in the multi-field case.
+    field = "Record_Type" if schema is not None and len(fields) == len(schema.fields) else fields[-1]
+
+    if error_category == ParserErrorCategoryChoices.PRE_CHECK:
```
and category 4 errors

elipe17 committed:
- include cat4 in record check

elipe17 commented:
Once all fra submission history code is merged, I will implement aggregates functionality

jtimpe commented on tdrs-backend/tdpservice/data_files/error_reports.py:
```diff
@@ -0,0 +1,341 @@
+    def get_row_generator(self):
+        """Get row generator for error report."""
+        return lambda error, exit_date, ssn, fields_json: (exit_date,
+                                                           self._obscure_ssn(ssn),
```
this (and the default `''` for `ssn`) results in the cat1/header error `File does not begin with FRA data` to also have a partial obscured ssn. suggest defaulting to `None` and `self._obscure_ssn(ssn) if ssn else ''`

jtimpe reviewed (approved):
tiny suggestion, otherwise works great

mattcoleanderson reviewed (approved):
lgtm!

elipe17 committed:
- Update submission history to generate total errors field

elipe17 committed:
- Add null check for ssn

elipe17 committed:
- linting

raftmsohani reviewed (approved):
LGTM!

elipe17 committed:
Merge branch '3527-parser-error-values' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

lhuxraft reviewed (changes requested):
@elipe17 Testing now and noticing a few issues:
1. When testing with [TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19451582/TEST_FRA_CSV.csv), I get the following error report (which I don't think is accurate) because I'm using the Exit Date / SSN format. Can you confirm if this is expected behavior?
![Image](https://github.com/user-attachments/assets/b18129e8-34c7-4d0c-8adc-c809751eb10f)
2. I also tested [TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19451632/TEST_FRA_XLSX.xlsx) and got 
[2024-Q3-Work Outcomes of TANF Exiters.xlsx](https://github.com/user-attachments/files/19451637/2024-Q3-Work.Outcomes.of.TANF.Exiters.xlsx) as my error report. A few issues with this error report:
- The error message language does not match the error message language documented in the report that is linked in this ticket. It looks like it's using the TANF/SSP error language, which it should not. Attaching the report here again for reference: [Updated Error Report (5) (3).xlsx](https://github.com/user-attachments/files/19451687/Updated.Error.Report.5.3.xlsx)
- This error report should also only be checking for the types of errors listed in the report I linked above. Again, it looks like it might be leveraging TANF/SSP validation rules (there should be only be three total errors that a user receives, as documented in the 'Updated Error Report' file).

Can you make these changes and then I'll re-review?

elipe17 commented:
> @elipe17 Testing now and noticing a few issues:
> 
> 1. When testing with [TEST_FRA_CSV.csv](https://github.com/user-attachments/files/19451582/TEST_FRA_CSV.csv), I get the following error report (which I don't think is accurate) because I'm using the Exit Date / SSN format. Can you confirm if this is expected behavior?
>    ![Image](https://private-user-images.githubusercontent.com/187318776/426615767-b18129e8-34c7-4d0c-8adc-c809751eb10f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI5MTk1NDEsIm5iZiI6MTc0MjkxOTI0MSwicGF0aCI6Ii8xODczMTg3NzYvNDI2NjE1NzY3LWIxODEyOWU4LTM0YzctNGQwYy04YWRjLWM4MDk3NTFlYjEwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNVQxNjE0MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMWMxMGE0Y2ExM2U1MDJjMDVjYjA1Yzg2NGJlNGU2NDczMzBjMjg5MWY2NTk3ZTU3N2E1YzA1MjI2MzQwMTRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.H6cKw6kxVRyENZ0eWnCm50lB6PrZ5NR8guS0deZp3_s)
> 2. I also tested [TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/19451632/TEST_FRA_XLSX.xlsx) and got
>    [2024-Q3-Work Outcomes of TANF Exiters.xlsx](https://github.com/user-attachments/files/19451637/2024-Q3-Work.Outcomes.of.TANF.Exiters.xlsx) as my error report. A few issues with this error report:
> 
> * The error message language does not match the error message language documented in the report that is linked in this ticket. It looks like it's using the TANF/SSP error language, which it should not. Attaching the report here again for reference: [Updated Error Report (5) (3).xlsx](https://github.com/user-attachments/files/19451687/Updated.Error.Report.5.3.xlsx)
> * This error report should also only be checking for the types of errors listed in the report I linked above. Again, it looks like it might be leveraging TANF/SSP validation rules (there should be only be three total errors that a user receives, as documented in the 'Updated Error Report' file).
> 
> Can you make these changes and then I'll re-review?

@lhuxraft 

- Your CSV file begins with a BOM which is why that one is failing. You must submit ASCII or UTF-8 (no BOM) csv files.
- The error report for the xlsx seems right since you submitted the file for 2024 Q3. Most of the errors are because your test file does not conform to the submission date you chose.
- The language is determined based on the rules in the screenshot below. We have some wiggle room wrt some of the language. However, we cannot generalize messages because the message is generated based on the failing validator. At runtime we don't "know" what the failing validator is, we just get it's error message.
- If you look [here](https://github.com/raft-tech/TANF-app/blob/3419-fra-error-report/tdrs-backend/tdpservice/parsers/schema_defs/fra/te1.py), you will see that all checks the screenshot below mentions are implemented.

![Screenshot 2025-03-25 at 12 23 37 PM](https://github.com/user-attachments/assets/9105a632-893d-4c32-bf1a-77fc48cd5e37)

If you have any other questions i'd be happy to hop on a call and discuss.

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
- Update te1 to aggregate field validators
- Update fra validation logic to not perform field validation if pre parsing validation fails
- Update FRA error messages to be aggregated based on validators
- updated tests

elipe17 committed:
Merge branch 'develop' into 3419-fra-error-report

elipe17 commented:
@lhuxraft I updated this PR so that it gets as close as possible to the expected error report format.

lhuxraft commented:
> @lhuxraft I updated this PR so that it gets as close as possible to the expected error report format.

Thank you, I really appreciate it! Will review again later this afternoon

lhuxraft commented:
@elipe17 This is looking great - thanks for making the requested updates! Testing again and noticing a couple things:

1. The error messages look great! However, I am noticing for the 'Error Location in File' column that the row letter is sometimes cut off - could you look into this? See screenshot:
![image](https://github.com/user-attachments/assets/c4edcc07-1a8f-4fe7-b24d-b15f0c400955)
2. I am also noticing that I am not receiving a duplicate error when I have the same SSN in a month - example file I was using: [FRA_TEST_DUPE.csv](https://github.com/user-attachments/files/19474562/FRA_TEST_DUPE.csv)
3. This is a nit, but can we capitalize SSN in the error report's column naming?

elipe17 commented:
> @elipe17 This is looking great - thanks for making the requested updates! Testing again and noticing a couple things:
> 
> 1. The error messages look great! However, I am noticing for the 'Error Location in File' column that the row letter is sometimes cut off - could you look into this? See screenshot:
>    ![image](https://private-user-images.githubusercontent.com/187318776/427199705-c4edcc07-1a8f-4fe7-b24d-b15f0c400955.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMTg5OTYsIm5iZiI6MTc0MzAxODY5NiwicGF0aCI6Ii8xODczMTg3NzYvNDI3MTk5NzA1LWM0ZWRjYzA3LTFhOGYtNGZlNy1iMjRkLWIxNWYwYzQwMDk1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNlQxOTUxMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MTE1NTc2ODYzMjQ5YjM3MTk3ZTY4MDNjY2JmNGYyMjA0ZDQzNzg2NDNlZjdlM2RlMGMzNDY5ODg3ZjZmMTdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.t3rWywl57g7CiOPVAyBAlHJveb06OBRl4_VBymtv1aY)
> 2. I am also noticing that I am not receiving a duplicate error when I have the same SSN in a month - example file I was using: [FRA_TEST_DUPE.csv](https://github.com/user-attachments/files/19474562/FRA_TEST_DUPE.csv)
> 3. This is a nit, but can we capitalize SSN in the error report's column naming?

- The missing column index isn't actually missing. This is a side effect of running the submission date check against the Exit date. It runs as a preparsing validator which indicates that the entire row of data was invalid if it's check failed. That's why you only see a number (the row) for the rows in the report with that error. That is not something we can change without a big refactor since TANF relies on that style of functionality too
- Dup checking isn't implemented on this PR. That lives in the cat4 PR since it is a true cat4 error.
- I can make that change

Edit: 
SSN is now flaming caps in report
![Screenshot 2025-03-26 at 4 14 29 PM](https://github.com/user-attachments/assets/3b736dce-c444-4e50-9d58-64464c6d84dc)

elipe17 committed:
- Update so that ssn is all caps

lhuxraft commented:
> > @elipe17 This is looking great - thanks for making the requested updates! Testing again and noticing a couple things:
> > 
> > 1. The error messages look great! However, I am noticing for the 'Error Location in File' column that the row letter is sometimes cut off - could you look into this? See screenshot:
> >    ![image](https://private-user-images.githubusercontent.com/187318776/427199705-c4edcc07-1a8f-4fe7-b24d-b15f0c400955.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMTg5OTYsIm5iZiI6MTc0MzAxODY5NiwicGF0aCI6Ii8xODczMTg3NzYvNDI3MTk5NzA1LWM0ZWRjYzA3LTFhOGYtNGZlNy1iMjRkLWIxNWYwYzQwMDk1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNlQxOTUxMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MTE1NTc2ODYzMjQ5YjM3MTk3ZTY4MDNjY2JmNGYyMjA0ZDQzNzg2NDNlZjdlM2RlMGMzNDY5ODg3ZjZmMTdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.t3rWywl57g7CiOPVAyBAlHJveb06OBRl4_VBymtv1aY)
> > 2. I am also noticing that I am not receiving a duplicate error when I have the same SSN in a month - example file I was using: [FRA_TEST_DUPE.csv](https://github.com/user-attachments/files/19474562/FRA_TEST_DUPE.csv)
> > 3. This is a nit, but can we capitalize SSN in the error report's column naming?
> 
> * The missing column index isn't actually missing. This is a side effect of running the submission date check against the Exit date. It runs as a preparsing validator which indicates that the entire row of data was invalid if it's check failed. That's why you only see a number (the row) for the rows in the report with that error. That is not something we can change without a big refactor since TANF relies on that style of functionality too
> * Dup checking isn't implemented on this PR. That lives in the cat4 PR since it is a true cat4 error.
> * I can make that change
> 
> Edit: SSN is now flaming caps in report ![Screenshot 2025-03-26 at 4 14 29 PM](https://private-user-images.githubusercontent.com/125676261/427212265-3b736dce-c444-4e50-9d58-64464c6d84dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMjE5NjAsIm5iZiI6MTc0MzAyMTY2MCwicGF0aCI6Ii8xMjU2NzYyNjEvNDI3MjEyMjY1LTNiNzM2ZGNlLWM0NDQtNGU1MC05ZDU4LTY0NDY0YzZkODRkYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNlQyMDQxMDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNmUyZTk0Yjg5MzZmYjQyODhjZDhkYTg1NjMzMzc4ZjkxOTE1ZWVlM2RiYTk1ZGMwZDliMDg5NDRjYTA3ZGIxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.q2GJrSWx6ExuK-hve2AmGR1okRIMLLGAEG9m2KeaaTc)

@elipe17 Thank you for explaining! Just spoke with UX and would it be easier from a technical standpoint to have that column display row number instead (removing the 'B' entirely)? We would also want to change the column name to 'Row Number in File' to accommodate that change

elipe17 commented:
> > > @elipe17 This is looking great - thanks for making the requested updates! Testing again and noticing a couple things:
> > > 
> > > 1. The error messages look great! However, I am noticing for the 'Error Location in File' column that the row letter is sometimes cut off - could you look into this? See screenshot:
> > >    ![image](https://private-user-images.githubusercontent.com/187318776/427199705-c4edcc07-1a8f-4fe7-b24d-b15f0c400955.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMTg5OTYsIm5iZiI6MTc0MzAxODY5NiwicGF0aCI6Ii8xODczMTg3NzYvNDI3MTk5NzA1LWM0ZWRjYzA3LTFhOGYtNGZlNy1iMjRkLWIxNWYwYzQwMDk1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNlQxOTUxMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MTE1NTc2ODYzMjQ5YjM3MTk3ZTY4MDNjY2JmNGYyMjA0ZDQzNzg2NDNlZjdlM2RlMGMzNDY5ODg3ZjZmMTdjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.t3rWywl57g7CiOPVAyBAlHJveb06OBRl4_VBymtv1aY)
> > > 2. I am also noticing that I am not receiving a duplicate error when I have the same SSN in a month - example file I was using: [FRA_TEST_DUPE.csv](https://github.com/user-attachments/files/19474562/FRA_TEST_DUPE.csv)
> > > 3. This is a nit, but can we capitalize SSN in the error report's column naming?
> > 
> > 
> > 
> > * The missing column index isn't actually missing. This is a side effect of running the submission date check against the Exit date. It runs as a preparsing validator which indicates that the entire row of data was invalid if it's check failed. That's why you only see a number (the row) for the rows in the report with that error. That is not something we can change without a big refactor since TANF relies on that style of functionality too
> > * Dup checking isn't implemented on this PR. That lives in the cat4 PR since it is a true cat4 error.
> > * I can make that change
> > 
> > Edit: SSN is now flaming caps in report ![Screenshot 2025-03-26 at 4 14 29 PM](https://private-user-images.githubusercontent.com/125676261/427212265-3b736dce-c444-4e50-9d58-64464c6d84dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMjE5NjAsIm5iZiI6MTc0MzAyMTY2MCwicGF0aCI6Ii8xMjU2NzYyNjEvNDI3MjEyMjY1LTNiNzM2ZGNlLWM0NDQtNGU1MC05ZDU4LTY0NDY0YzZkODRkYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyNlQyMDQxMDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNmUyZTk0Yjg5MzZmYjQyODhjZDhkYTg1NjMzMzc4ZjkxOTE1ZWVlM2RiYTk1ZGMwZDliMDg5NDRjYTA3ZGIxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.q2GJrSWx6ExuK-hve2AmGR1okRIMLLGAEG9m2KeaaTc)
> 
> @elipe17 Thank you for explaining! Just spoke with UX and would it be easier from a technical standpoint to have that column display row number instead (removing the 'B' entirely)? We would also want to change the column name to 'Row Number in File' to accommodate that change

@lhuxraft that's an easy change if we want to go that route. Then we don't have the two types of locations in the column.

Edit:
Here is what that would look like.

![Screenshot 2025-03-27 at 8 07 26 AM](https://github.com/user-attachments/assets/9e47c3fa-2498-49d1-b2e8-79a949f8c7a6)

elipe17 committed:
- Updated error location so that it is only row based
- Autofit after report is done

elipe17 committed:
- Update tests

elipe17 committed:
Merge branch 'develop' into 3419-fra-error-report

victoriaatraft reviewed (approved):
LGTM from testing and just left some visual feedback on #3419

lhuxraft commented:
Thank you @elipe17 for the changes! Did a quick re-review and LGTM 🚀 

Going to go ahead and move to QASP, but have one final minor comment:
- When I uploaded a file with just Exit Date errors, I get a row at the bottom that says '0' and 'No records created' - I don't believe this is intentional? (see screenshot). FWIW, I do not see this on all files that I submit, just for this one use case.
![image](https://github.com/user-attachments/assets/4f730c34-1fe3-4507-8db1-36eef285052d)

ADPennington committed:
Merge branch 'develop' into 3419-fra-error-report

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3419-fra-error-report

elipe17 committed:
- Update error generator logic
- Remove reference to document in error generator

lhuxraft commented:
@ADPennington Unblocked now - removing the Blocked label

elipe17 committed:
Merge branch 'develop' into 3419-fra-error-report

elipe17 committed:
Merge branch 'develop' into 3419-fra-error-report
